app/Http/Livewire/ModuleContent.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+namespace App\Http\Livewire;
+
+use App\Models\Module;
+use App\Models\User;
+use Illuminate\Support\Facades\Auth;
+use Livewire\Component;
+
+class ModuleContent extends Component
+{
+    public Module $module;
+
+    public User $user;
+
+    public ?string $startedContentAt = null;
+
+    public ?string $finishedContentAt = null;
+
+    public function mount(Module $module)
+    {
+        $this->module = $module;
+        $this->user = Auth::user();
+        $moduleUser = $this->user->modules->find($module->id);
+        $this->startedContentAt = $moduleUser?->getRelationValue('pivot')->started_content_at;
+        $this->finishedContentAt = $moduleUser?->getRelationValue('pivot')->finished_content_at;
+    }
+
+    protected $listeners = ['onPlayerEnd', 'onPlayerStart'];
+
+    public function onPlayerEnd()
+    {
+        if (! $this->finishedContentAt) {
+            $this->user->modules()->updateExistingPivot(
+                $this->module->id, ['finished_content_at' => now()]
+            );
+            $this->finishedContentAt = 'finished_content';
+            $finishedCourse = true;
+            $course = $this->module->course;
+            foreach ($course->modules as $module) {
+                $moduleUser = $module->users->find($this->user->id)?->getRelationValue('pivot');
+                if (! $moduleUser) {
+                    $finishedCourse = false;
+                    break;
+                } elseif (! $moduleUser->finished_content_at) {
+                    $finishedCourse = false;
+                    break;
+                }
+            }
+            if ($finishedCourse) {
+                $course->users()->updateExistingPivot(
+                    $this->user->id, ['finished_at' => now()]
+                );
+            }
+        }
+    }
+
+    public function onPlayerStart()
+    {
+        if (! $this->startedContentAt) {
+            $course = $this->module->course()->first();
```

Comment:
A module only has one course, so I think you could write this as:

```suggestion
            $course = $this->module->course;
```